server/src/main/java/org/elasticsearch/cluster/ClusterModule.java

```diff
@@ -340,23 +340,24 @@ public static Collection<AllocationDecider> createAllocationDeciders(
     ) {
         // collect deciders by class so that we can detect duplicates
         Map<Class<?>, AllocationDecider> deciders = new LinkedHashMap<>();
-        addAllocationDecider(deciders, new MaxRetryAllocationDecider());
-        addAllocationDecider(deciders, new ResizeAllocationDecider());
+        // computationally cheaper deciders should be ordered first to skip more expensive ones in case of NO or THROTTLE decision
+        addAllocationDecider(deciders, new EnableAllocationDecider(clusterSettings));
+        addAllocationDecider(deciders, new SameShardAllocationDecider(clusterSettings));
         addAllocationDecider(deciders, new ReplicaAfterPrimaryActiveAllocationDecider());
+        addAllocationDecider(deciders, new MaxRetryAllocationDecider());
+        addAllocationDecider(deciders, new ShardsLimitAllocationDecider(clusterSettings));
```

Comment:
I don't think this one is all that cheap (especially given that it uses streams)